include/circt/Dialect/HW/HWOpInterfaces.td

```diff
@@ -32,79 +32,67 @@ def HWModuleLike : OpInterface<"HWModuleLike", [
   let description = "Provide common module information.";
 
   let methods = [
-    InterfaceMethod<"Get the number of ports",
-    "size_t", "getNumPorts">,
 
     InterfaceMethod<"Get a port symbol attribute",
     "::circt::hw::InnerSymAttr", "getPortSymbolAttr", (ins "size_t":$portIndex)>,
 
     InterfaceMethod<"Get the module type",
-    "::circt::hw::ModuleType", "getHWModuleType", (ins),
-    /*methodBody=*/[{}],
-    /*defaultImplementation=*/[{ return ::circt::hw::detail::fnToMod($_op, getInputNames(), getOutputNames()); }]>,
-
-    InterfaceMethod<"Return the names of the inputs to this module",
-    "mlir::ArrayAttr", "getInputNames", (ins),
-    /*methodBody=*/[{}],
-    /*defaultImplementation=*/[{
-      return $_op->template getAttrOfType<ArrayAttr>("argNames");
-    }]>,
+    "::circt::hw::ModuleType", "getHWModuleType", (ins)>,
+  ];
 
-    InterfaceMethod<"Return the names of the outputs this module",
-    "mlir::ArrayAttr", "getOutputNames", (ins),
-    /*methodBody=*/[{}],
-    /*defaultImplementation=*/[{
-      return $_op->template getAttrOfType<ArrayAttr>("resultNames");
-    }]>,
+  let extraSharedClassDeclaration = [{
 
-    InterfaceMethod<"Return the locations of the inputs to this module",
-    "mlir::ArrayAttr", "getInputLocs", (ins),
-    /*methodBody=*/[{}],
-    /*defaultImplementation=*/[{
-      return $_op->template getAttrOfType<ArrayAttr>("argLocs");
-    }]>,
+    size_t getNumPorts() {
+      return $_op.getHWModuleType().getPorts().size();
+    }
 
-    InterfaceMethod<"Return the locations of the outputs of this module",
-    "mlir::ArrayAttr", "getOutputLocs", (ins),
-    /*methodBody=*/[{}],
-    /*defaultImplementation=*/[{
-      return $_op->template getAttrOfType<ArrayAttr>("resultLocs");
-    }]>,
+    size_t getNumInputs() {
+      return $_op.getHWModuleType().getNumInputs();
+    }
 
-  ];
+    size_t getNumOutputs() {
+      return $_op.getHWModuleType().getNumOutputs();
+    }
 
-  let extraSharedClassDeclaration = [{
+    SmallVector<StringAttr> getInputNames() {
+      return $_op.getHWModuleType().getInputNames();
+    }
 
-    // Return the number of inputs to this module
-    unsigned getNumInputs() {
-      return getInputNames().size();
+    SmallVector<StringAttr> getOutputNames() {
+      return $_op.getHWModuleType().getOutputNames();
     }
 
-    // Return the number of outputs from this module
-    unsigned getNumOutputs() {
-      return getOutputNames().size();
+    ArrayAttr getInputNamesAttr() {
+      auto ty = $_op.getHWModuleType();
+      auto names = ty.getInputNames();
+      SmallVector<Attribute> names_(names.begin(), names.end());
+      return ArrayAttr::get(ty.getContext(), names_);
```

Comment:
Really, we have to copy to convert `<StringAttr>` to `<Attribute>`? :disappointed: . Boo!

Something something array attr taking a map_range or something :/.